### PR TITLE
Initial version, leverage rust for hid transfer

### PR DESCRIPTION
**Pull request by vcdmp**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [ ] Feature
- [ ] Documentation

[fwupdtool_vv_get-devices.zip](https://github.com/user-attachments/files/17298938/fwupdtool_vv_get-devices.zip)

### CODE REVIEW

File: plugins/logitech-tap/fu-logitech-tap.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ }
+ 
+ #[derive(New)]
+ struct FuStructLogitechTapSensorHidReq {
+     cmd: FuStructLogitechTapSensorHidSetCmd,
+      //payload goes here
+      payload_byte1: u8,
+      payload_byte2: u8,
+      payload_byte3: u8 == 0x,
+      payload_byte4: u8 == 0x,
```
Comment (hughsie) [inline plugins/logitech-tap/fu-logitech-tap.rs:29]: i think you want `0x0` here

File: plugins/logitech-tap/fu-logitech-tap-sensor-device.c [hunk 2]
```diff
@@ -312,6 +312,7 @@ fu_logitech_tap_sensor_device_init(FuLogitechTapSensorDevice *self)
 	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_TRIPLET);
 	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
 	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_SIGNED_PAYLOAD);
+	fu_device_add_vendor_id(FU_DEVICE(self), "USB:0x0872");
```
Comment (hughsie) [inline plugins/logitech-tap/fu-logitech-tap-sensor-device.c:315]: I think this is not needed when https://github.com/fwupd/fwupd/pull/7762 is merged.

File: plugins/logitech-tap/fu-logitech-tap-sensor-device.c [hunk 3]
```diff
@@ -90,16 +91,15 @@ static gboolean
...
-	    kHidMcuTdeReportId,
-	    kHidMcuTdeModeSelector,
-	    kHidMcuTdeModeEnable,
-	    0,
-	    0,
-	};
+	g_autoptr(FuStructLogitechTapSensorHidReq) st = fu_struct_logitech_tap_sensor_hid_req_new();
+	fu_struct_logitech_tap_sensor_hid_req_set_cmd(st,
+						      FU_STRUCT_LOGITECH_TAP_TOUCH_HID_SET_CMD_TDE);
+	fu_struct_logitech_tap_sensor_hid_req_set_payload_byte1(st, kHidMcuTdeModeSelector);
```
Comment (hughsie) [inline plugins/logitech-tap/fu-logitech-tap-sensor-device.c:97]: I think moving kHidMcuTde* to rs format will give quite a bit of type safety too.

File: plugins/logitech-tap/fu-logitech-tap.rs [hunk 4]
```diff
@@ -34,3 +34,68 @@ struct FuStructLogitechTapTouchHidReq {
...
+ }
+
+ #[derive(New)]
+ struct FuStructLogitechTapSensorHidRes {
+     cmd: FuLogitechTapSensorHidGetCmd,
+      //payload goes here
+      payload_byte1: u8 = 0x0,
+      payload_byte2: u8 = 0x0,
+      payload_byte3: u8 = 0x0,
+      payload_byte4: u8 = 0x0,
```
Comment (hughsie) [inline plugins/logitech-tap/fu-logitech-tap.rs:100]: you don't need the = 0x0 here -- it's memset to zero by default. Then squash down and rebase and I think it's good -- thanks!

Comment (vcdmp) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: Unfortunately,  we are hitting the same regression issue again (No detected devices) with this video4linux/hidraw device.

Everything works fine when we use 1.9.24 codebase. But same device, on same platform/setup fails, with latest  codebase.

vv logs not helping much. Plugin is getting invoked but not device_init()/device_class_init(). Any pointers!

Comment (hughsie) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: > with latest codebase

Do you mean the `main` branch or `wip/vcdmp/tap_sensor_rust_refactor_v0`? Can you share the -vv logs again please on the failing branch. Thanks.

Comment (vcdmp) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: seeing regression with main branch. vv logs attached.
[fwupdtool_vv.zip](https://github.com/user-attachments/files/17282556/fwupdtool_vv.zip)

Comment (hughsie) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: > 17:36:17.854 FuEngine             failed to setup backend udev: failed to set up netlink: bind to udev socket failed: Address already in use

Your build directory is linking in old libraries. Do something like `mv venv venv.old` and then re-do `./contrib/setup` and it should work again.

Comment (hughsie) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: > $ fwupdtool --plugins logitech_tap get-devices --verbose --verbose

You're not using sudo? Does it work if you use that? Did it work before if you didn't use sudo?

Comment (superm1) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: > You're not using sudo? Does it work if you use that? Did it work before if you didn't use sudo?

`venv` does it for you.

Comment (hughsie) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: @superm1 for the `failed to set up netlink: bind to udev socket failed: Address already in use` is there any in-process tool to report *what* is binding to the same socket? I've not seen this in a loong time, and there shouldn't be anything linking against gudev at all in fwupd now.

Comment (superm1) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: Well so I saw the exact same thing on one of my machines... IIRC rebooted it and it _seemed_ to recover.  But knowing that it's happening to another person I'm not so sure what could be going on.

Comment (hughsie) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: I wonder if venv is affecting the linking -- i.e. does `sudo ./src/fwupdtool get-devices` also do the same?

Comment (hughsie) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: @vcdmp if you do `sudo gdb --args ./src/fwupdtool --plugins logitech_tap get-devices --verbose --verbose` and then `b device_monitor_new_full` then `r` -- do you get a breakpoint stop? If so, what does `bt` say when the breakpoint has been hit?

Comment (superm1) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: Yeah I'd say axe the whole `venv` directory to rule it out.  `./contrib/setup` will rebuild it.

Comment (vcdmp) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: Did I missed step here? I tried both, within venv and without venv"
ubuntu22@ubuntu22-ThinkPad-P1:~/repo/fwupd$ v4l2-ctl --list-devices
Synaptics RMI4 Touch Sensor (rmi4:rmi4-00.fn54):
	/dev/v4l-touch0

Logitech Tap HDMI Capture (usb-0000:00:14.0-1.2):
	/dev/video4
	/dev/video5
	/dev/media2

Integrated IR Camera: Integrate (usb-0000:00:14.0-12):
	/dev/video2
	/dev/video3
	/dev/media1

Integrated Camera: Integrated C (usb-0000:00:14.0-8):
	/dev/video0
	/dev/video1
	/dev/media0

ubuntu22@ubuntu22-ThinkPad-P1:~/repo/fwupd$ sudo gdb --args ./src/fwupdtool --plugins logitech_tap get-devices --verbose --verbose
[sudo] password for ubuntu22: 
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04.2) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
./src/fwupdtool: No such file or directory.
(gdb) b device_monitor_new_full
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (device_monitor_new_full) pending.
(gdb) r --
Starting program:  --
No executable file specified.
Use the "file" or "exec-file" command.
(gdb)

Comment (vcdmp) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: Usually I do "DEBUG=1 fwupdtoo...." for live debugging

Comment (vcdmp) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: FYI, observing same issue on last release branch "2.0.0".

Comment (hughsie) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: >  ./src/fwupdtool: No such file or directory.

I think you need to do `cd venv/build` to run that directly. 

> Usually I do "DEBUG=1 fwupdtoo...." for live debugging

If that runs gdb, that's fine too. I need the `bt` to see what's linking in udev. Thanks.

Comment (hughsie) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: In the meantime, to unblock your testing, can you try the fix in https://github.com/fwupd/fwupd/pull/7869 i.e. just cherry pick 94edef0b74177fcf16743608e80ceb8f30404e00 please.

Comment (vcdmp) [reply plugins/logitech-tap/fu-logitech-tap.rs:100]: @hughsie Recloned main branch to pickup your recent [changes](https://github.com/fwupd/fwupd/pull/7869).  

Seems like it is working when invoking fwupdtool directly from build folder.
Same setup fails with "No detected devices" error, when fwupdtool invoked inside virtual environment. 
I have attached vv logs for both usecases.

[fwupdtool_vv.zip](https://github.com/user-attachments/files/17297014/fwupdtool_vv.zip)

**Comment by superm1** [comment]
What Ubuntu version are you on?

**Comment by superm1** [comment]
> What Ubuntu version are you on?

I think there is a problem with venv on Ubuntu 22.04 but not on Ubutnu 24.04.  I don't really understand why right now.

**Comment by superm1** [comment]
Can you please rebase and squash to one commit?

**Comment by vcdmp** [comment]
> > What Ubuntu version are you on?
> 
> I think there is a problem with venv on Ubuntu 22.04 but not on Ubuntu 24.04. I don't really understand why right now.

$ uname -a
Linux ubuntu22-ThinkPad-P1 6.8.0-45-generic #45~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Sep 11 15:25:05 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

**Comment by superm1** [comment]
> 
> 
> 
> > > What Ubuntu version are you on?
> > 
> > I think there is a problem with venv on Ubuntu 22.04 but not on Ubuntu 24.04. I don't really understand why right now.
> 
> $ uname -a
> Linux ubuntu22-ThinkPad-P1 6.8.0-45-generic #45~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Sep 11 15:25:05 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
> 

Ok that's consistent with what I saw then, I remember I was trying a few machines with main and I had an old one hitting this too. When I hit it I just used another machine.

I'm not sure why outside venv works tho.

**Comment by superm1** [comment]
> Can you please rebase and squash to one commit?

It seems you didn't rebase on origin/main. Try one more time and squash all commits when you did it.

**Comment by vcdmp** [comment]
> > Can you please rebase and squash to one commit?
> 
> It seems you didn't rebase on origin/main. Try one more time and squash all commits when you did it.

I am not that used to with squash/merge. Last time you told me to run following, I just did it. What is the next step!!!:

ubuntu22@ubuntu22-ThinkPad-P1:~/repo/fwupd$ git rebase -i origin/main
Successfully rebased and updated refs/heads/wip/vcdmp/tap_sensor_rust_refactor_v0.
ubuntu22@ubuntu22-ThinkPad-P1:~/repo/fwupd$ git push --force origin
Everything up-to-date
ubuntu22@ubuntu22-ThinkPad-P1:~/repo/fwupd$

**Review by superm1 — APPROVED** [review]

**Pull request merged**